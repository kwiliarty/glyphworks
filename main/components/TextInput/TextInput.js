import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components' 

const StyledDiv = styled.div`
  display: flex;
  flex-direction: column;
  font-size: ${ props => props.theme.smallFontSize };
`

const StyledLabel = styled.label`
  margin-bottom: 0.1em;
`

const StyledInputLine = styled.div`
  display: flex;
  align-items: center;
`

const StyledInput = styled.input`
  flex-grow: 1;
  border: 1px solid ${ props => props.theme.mainTextColor };
  border-radius: ${ props => props.theme.borderRadius };
  font-family: ${ props => props.theme.ipaFontFamily };
  font-size: ${ props => props.theme.baseFontSize };
  color: ${ props => props.theme.mainTextColor };
  text-indent: 0.3em;
  &:focus {
    outline: 2px solid ${ props => props.theme.mainTextColor };
  }
`

const StyledSuffix = styled.div`
`

const TextInput = props => {

  const idStem = ( props.idStem || Math.random().toString( 36 ).substr( 2, 16 ) )
  const { children } = props

  return (
    <StyledDiv { ...props.wrapperProps }>
      <StyledLabel htmlFor={ `id-${idStem}` }>{ props.labelText }</StyledLabel>
      <StyledInputLine>
        <StyledInput
          { ...props.inputProps }
          type='text'
          id={ `id-${idStem}` }
        />
        <StyledSuffix { ...props }>
          { children && [].concat( children ).filter( child => child.type.nickname == 'Suffix' ) }
        </StyledSuffix>
      </StyledInputLine>
    </StyledDiv>
  )
}

const Suffix = props => props.children
Suffix.nickname = 'Suffix'
TextInput.Suffix = Suffix

TextInput.propTypes = {
  /** This component displays children of its subcomponents. */
  children: PropTypes.node,
  /** An optional id stem for the input fields. If none is provided, we create a random one. */
  idStem: PropTypes.string,
  /** Properties that will be passed to the `<input>` element. */
  inputProps: PropTypes.object,
  /** Label for the text input */
  labelText: PropTypes.string.isRequired,
  /** Properties that will be passed into the wrapper <div>. */
  wrapperProps: PropTypes.object,
}

TextInput.defaultProps = {
  inputProps: {},
  wrapperProps: {},
}

export default TextInput
