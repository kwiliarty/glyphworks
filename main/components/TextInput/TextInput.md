A generic text input.

```jsx
<TextInput
  labelText='My Label'
/>
```

You can specify a suffix for the input field:

```jsx
<TextInput
  labelText='I have a suffix'
>
  <TextInput.Suffix>
    <div style={{ paddingLeft: '0.5em' }}>Suffix!</div>
  </TextInput.Suffix>
</TextInput>
```
