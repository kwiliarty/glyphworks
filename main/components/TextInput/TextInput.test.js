import React from 'react'
import { render, screen } from '@testing-library/react'
import TextInput from './TextInput.js'

describe( 'The TextInput component', () => {

  it( 'renders an input with a label', () => {
    render( <TextInput labelText='Testing' idStem='test' /> )
    const input = screen.getByRole( 'textbox' )
    expect( input ).toMatchSnapshot()
  })
})
