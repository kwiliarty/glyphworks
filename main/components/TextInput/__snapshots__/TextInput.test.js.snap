// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The TextInput component renders an input with a label 1`] = `
.c0 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  border: 1px solid #625418;
  border-radius: 0.1em;
  font-family: GentiumPlus,serif;
  font-size: 28px;
  color: #625418;
  text-indent: 0.3em;
}

.c0:focus {
  outline: 2px solid #625418;
}

<input
  class="c0"
  id="id-test"
  type="text"
/>
`;
